xds/internal/xdsclient/authority.go

```diff
@@ -0,0 +1,229 @@
+func (c *clientImpl) findAuthority(n *xdsresource.Name) (retA *authority, _ error) {
+	scheme, authority := n.Scheme, n.Authority
+
+	c.authorityMu.Lock()
+	defer c.authorityMu.Unlock()
+	if c.done.HasFired() {
+		return nil, errors.New("the xds-client is closed")
+	}
+
+	defer func() {
+		// All returned authority from this function will be used by a watch,
+		// holding the ref here.
+		//
+		// Note that this must be done while c.authorityMu is held, to avoid the
+		// race that an authority is returned, but before the watch starts, the
+		// old last watch is canceled (in another goroutine), causing this
+		// authority to be removed, and then a watch will start on a removed
+		// authority.
+		//
+		// unref() will be done when the watch is canceled.
+		if retA != nil {
+			retA.ref()
+		}
+	}()
+
+	var config *bootstrap.ServerConfig
+	if scheme != federationScheme {
+		config = c.config.XDSServer
+	} else {
+		authConfig, ok := c.config.Authorities[authority]
+		if !ok {
+			return nil, fmt.Errorf("xds: failed to find authority %q", authority)
+		}
+		config = authConfig.XDSServer
+	}
+
+	a, err := c.newAuthority(config)
+	if err != nil {
+		return nil, fmt.Errorf("xds: failed to connect to the control plane for authority %q: %v", authority, err)
+	}
+	return a, nil
+}
+
+// newAuthority creates a new authority for the config. But before that, it
+// checks the cache to see if an authority for this config already exists.
+//
+// caller must hold c.authorityMu
+func (c *clientImpl) newAuthority(config *bootstrap.ServerConfig) (_ *authority, retErr error) {
+	// First check if there's already an authority for this config. If found, it
+	// means there was another watch with a different authority name but the
+	// same server config. Return it.
+	configStr := config.String()
+	if a, ok := c.authorities[configStr]; ok {
+		return a, nil
+	}
+	// Second check if there's an authority in the idle cache. If found, it
+	// means this authority was created, but moved to the idle cache because the
+	// watch was canceled. Move it from idle cache to the authority cache, and
+	// return.
+	if old, ok := c.idleAuthorities.Remove(configStr); ok {
+		oldA, _ := old.(*authority)
+		if oldA != nil {
+			c.authorities[configStr] = oldA
+			return oldA, nil
+		}
+	}
+
+	// Make a new authority since there's no existing authority for this config.
+	ret := &authority{config: config, pubsub: pubsub.New(c.watchExpiryTimeout, c.logger)}
+	defer func() {
+		if retErr != nil {
+			ret.close()
+		}
+	}()
+	ctr, err := newController(config, ret.pubsub, c.updateValidator, c.logger)
+	if err != nil {
+		return nil, err
+	}
+	ret.controller = ctr
+	// Add it to the cache, so it will be reused.
+	c.authorities[configStr] = ret
+	return ret, nil
+}
+
+// unrefAuthority unrefs the authority. It also moves the authority to idle
+// cache if it's ref count is 0.
+//
+// Caller must not hold c.authorityMu.
+func (c *clientImpl) unrefAuthority(a *authority) {
```

Comment:
Can we instead return a `cancel` or `unref` func from `clientImpl.newAuthority` that callers should invoked instead of this separate method?

The advantage with that approach is that we would not need to store `config` inside of `authority` because as I see it now, the only reason for that seems to be to delete the `authority` from the `authorities` map.

Reply:
Updated.
Note that `unref` pairs with `findAuthority`, not `newAuthority`.